Units: add new unit for duration, it is optimized for displaying days, hours, minutes and seconds.

pabigot opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**: Provide better representation of extended durations

**Which issue(s) this PR fixes**:

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #8781 

**Special notes for your reviewer**: If this needs modifications in documentation please point me to the relevant file.

pabigot committed:
DateTime: add support to show durations in days hours minutes seconds

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=24175) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=24175) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=24175) it.</sub>

pabigot commented:
@torkelo is there something I can do to help this make progress?

Loader23 reviewed (approved):
Looks good and I really want to see this in grafana. :-)
Makes me sad that nothing happend in these two weeks :-(

mckn commented:
> Looks good and I really want to see this in grafana. :-)
> Makes me sad that nothing happend in these two weeks :-(

I will have a look at this one on Monday and help you move this forward.

pabigot commented:
The description of the PR references an issue that has since been closed; the replacement issue is #25830.

mckn commented on packages/grafana-data/src/valueFormats/categories.ts:
```diff
@@ -318,6 +319,7 @@ export const getCategories = (): ValueFormatCategory[] => [
       { name: 'duration (ms)', id: 'dtdurationms', fn: toDurationInMilliseconds },
       { name: 'duration (s)', id: 'dtdurations', fn: toDurationInSeconds },
       { name: 'duration (hh:mm:ss)', id: 'dthms', fn: toDurationInHoursMinutesSeconds },
+      { name: 'duration (dd d hh:mm:ss)', id: 'dtdhms', fn: toDurationInDaysHoursMinutesSeconds },
```
```suggestion
      { name: 'duration (dd _d_ hh:mm:ss)', id: 'dtdhms', fn: toDurationInDaysHoursMinutesSeconds },
```

Sorry for being a bit slow but what does the third `d` stand for?

pabigot replied:
In the name it's **d**uration **t**o **d**ay **h**our **m**in **s**ec = dtdhms, parallels  existing **d**uration **t**o **h**our **m**in **s**ec = dthms.

In the parentheses it means the character `d` because there's no styling available to distinguish it from the `dd` indicating the day value.  Could express it as `DD d HH:MM:SS` but that'd be inconsistent with all the others.

mckn replied:
ah, Ignore the comment above. It is d = day(s)

mckn replied:
@pabigot how do you want this to work? I would probably expect the format to be `(dd hh:mm:ss)` to be consistent with the other formats in that list.

If the duration is 5 days, 2 hours, 15 minutes and 20 seconds. Will we then write the following: `05 d 02:15:20`?

mckn replied:
Otherwise I would probably only have the following `(d hh:mm:ss)`

pabigot replied:
That's a possibility.  There's some bikeshedding in https://github.com/grafana/grafana/issues/8781#issuecomment-432631725 where `dd:hh:mm:ss` is ruled out (very difficult to comprehend quickly).

I prefer the explicit separating `d`, but could tolerate `5 02:15:20`.  I do find it harder to read without the separator.

There should be no leading zeros for the days field as it can be any size, and the prefix element is eliminated if the value would be zero.

It seems the unit test should capture days less than 10, and days more than 99, to make that clear.

mckn replied:
Sorry for being so unclear 🙈 It is totally my fault.

What I meant was basically the description of the duration option. So in the drop down might be changed to `d hh:mm:ss` to have the same format as the other ones. 

```suggestion
      { name: 'duration (d hh:mm:ss)', id: 'dtdhms', fn: toDurationInDaysHoursMinutesSeconds },
```

I totally agree with you that the formatted  duration value still should have the `5 d 02:15:20` as you suggested. It is much more readable.

mckn replied:
The reason why I am suggesting only one `d` in the beginning of the format is because to indicate that we won't append a 0 if the number of days is less then 10.

Basically following the same convention that we have in momentjs: https://momentjs.com/docs/#/displaying/format/ where e.g. M (Month) is formatted as 1 and MM (Month) is formatted as 01.

mckn replied:
And the reason why I am suggesting to remove the `d` that represents the day from the format in the description is because we don't have those kind of things for the other duration formats. They ads `ago`, `weeks`, `days`, `hours` and other units to the output without it being presented in the options drop down.

pabigot replied:
Thanks; that's all clear now.  I'll rework this later today.

mckn commented on packages/grafana-data/src/valueFormats/dateTimeFormatters.test.ts:
```diff
@@ -179,6 +180,26 @@ describe('duration', () => {
+  it('-1179811 dtdhms', () => {
+    const str = toDurationInDaysHoursMinutesSeconds(-1179811);
+    expect(formattedValueToString(str)).toBe('13 d 15:43:31 ago');
+  });
```
Maybe add a unit test that verifies the output when we will get single digit days.

mckn commented on packages/grafana-data/src/valueFormats/dateTimeFormatters.ts:
```diff
@@ -313,6 +313,24 @@ export function toDurationInHoursMinutesSeconds(size: number): FormattedValue {
   return { text: strings.join(':') };
 }
 
+export function toDurationInDaysHoursMinutesSeconds(size: number): FormattedValue {
```
I think we probably can improve this code a bit but it does follow the same style as the other ones so lets keep it as it is. Great job!

pabigot commented:
@mckn I believe the following cover the requested changes.  Please let me know if I've missed something.

* Rebased on current master
* Removed `dd ` from description text
* Added unit tests for boundary durations (24 h and 100 h) which includes a single-digit day, and for a multi-year time to confirm large day counts

mckn commented:
> @mckn I believe the following cover the requested changes. Please let me know if I've missed something.
> 
> * Rebased on current master
> * Removed `dd ` from description text
> * Added unit tests for boundary durations (24 h and 100 h) which includes a single-digit day, and for a multi-year time to confirm large day counts

Awesome! Thanks for the contribution.

mckn commented:
@pabigot not sure but for some reason the CI didn't trigger. Please merge master into this branch and push that change and hopefully that will trigger the build.
